pygem/cad/ffd.py

```diff
@@ -0,0 +1,353 @@
+                        h_geom_edge, a, b = \
+                            BRep_Tool_Curve(topods_Edge(bspline_tshape_edge))
+                        this_bspline_edge = \
+                            geomconvert_CurveToBSplineCurve(h_geom_edge)
+                        bspline_geom_edge = this_bspline_edge
+                        h_bspline_edge.Add(this_bspline_edge, self.tolerance)
+                    edgesCount += 1
+
+                    edge_explorer.Next()
+
+                bspline_geom_hedge = h_bspline_edge.BSplineCurve()
+                bspline_geom_edge = bspline_geom_hedge
+
+                # number of knots is enriched here: this can become a user
+                # prescribed parameter for the class
+                firstParam = bspline_geom_edge.FirstParameter()
+                lastParam = bspline_geom_edge.LastParameter()
+                for i in range(self.knotsToAdd):
+                    bspline_geom_edge.InsertKnot(firstParam+ \
+                               i*(lastParam-firstParam)/self.knotsToAdd, 1, \
+                               self.tolerance)
+                shapesList = TopTools_ListOfShape()
+                # openCascade object
+                occ_edge = bspline_geom_edge
+
+                # extract the Control Points of each face
+                n_poles = occ_edge.NbPoles()
+                control_polygon_coordinates = np.zeros(\
+                         shape=(n_poles, 3))
+                # cycle over the poles to get their coordinates. The idea here
+                # is to move poles coordinates to deform the curves
+                i = 0
+                for pole in range(n_poles):
+                    control_point_coordinates = occ_edge.Pole(pole + 1)
+                    control_polygon_coordinates[i, :] = \
+                           [control_point_coordinates.X(), \
+                            control_point_coordinates.Y(), \
+                            control_point_coordinates.Z()]
+                    i += 1
+
+                ## CURVES PHASE ############################################
+                src_pts = control_polygon_coordinates
+                new_pts = super().__call__(src_pts) # dont touch this line
+                # save here the `new_pts` into the shape
+                ## END CURVES ##############################################
+
+                i = 0
+                for pole in range(n_poles):
+                    control_point = gp_Pnt(new_pts[i, 0],
+                                           new_pts[i, 1],
+                                           new_pts[i, 2])
+                    occ_edge.SetPole(pole + 1, control_point)
+                    i += 1
+
+                modified_edge = BRepBuilderAPI_MakeEdge(occ_edge).Edge()
+                shapesList.Append(modified_edge)
+
+                wire_maker = BRepBuilderAPI_MakeWire()
+                wire_maker.Add(shapesList)
+                result_wire = wire_maker.Wire()
```

Comment:
Class private method (one underscore in front of the name e.g. `_foo()`). Describe input and output variables.